// react-native-country-picker-modal/lib/FlagButton.js
import React, { memo, useState, useEffect } from 'react';
import { TouchableOpacity, StyleSheet, View, Platform } from 'react-native';
import { Flag } from './Flag';
import { useContext } from './CountryContext';
import { CountryText } from './CountryText';
import { useTheme } from './CountryTheme';
const styles = StyleSheet.create({
    container: {
        alignItems: 'flex-start',
        justifyContent: 'flex-start'
    },
    containerWithEmoji: {
        marginTop: 0
    },
    containerWithoutEmoji: {
        marginTop: 5
    },
    flagWithSomethingContainer: {
        flexDirection: 'row',
        flexWrap: 'wrap',
        alignItems: 'center'
    },
    something: { fontSize: 16 }
});
const FlagText = (props) => (React.createElement(CountryText, Object.assign({}, props, { style: styles.something })));
const FlagWithSomething = memo(({ countryCode, withEmoji, withCountryNameButton, withCurrencyButton, withCallingCodeButton, withFlagButton, flagSize }) => {
    const { translation, getCountryNameAsync, getCountryCurrencyAsync, getCountryCallingCodeAsync } = useContext();
    const [state, setState] = useState({
        countryName: '',
        currency: '',
        callingCode: ''
    });
    const { countryName, currency, callingCode } = state;
    useEffect(() => {
        if (withCountryNameButton) {
            getCountryNameAsync(countryCode, translation)
                .then((countryName) => setState({ ...state, countryName }))
                .catch(console.error);
        }
        if (withCurrencyButton) {
            getCountryCurrencyAsync(countryCode)
                .then((currency) => setState({ ...state, currency }))
                .catch(console.error);
        }
        if (withCallingCodeButton) {
            getCountryCallingCodeAsync(countryCode)
                .then((callingCode) => setState({ ...state, callingCode }))
                .catch(console.error);
        }
    }, [withCountryNameButton, withCurrencyButton, withCallingCodeButton]);
    return (React.createElement(View, { style: styles.flagWithSomethingContainer },
        React.createElement(Flag, Object.assign({}, { withEmoji, countryCode, withFlagButton, flagSize: flagSize })),
        countryName ? React.createElement(FlagText, null, countryName + ' ') : null,
        currency ? React.createElement(FlagText, null, `(${currency}) `) : null,
        callingCode ? React.createElement(FlagText, null, `+${callingCode}`) : null));
});
export const FlagButton = ({ withEmoji, withCountryNameButton, withCallingCodeButton, withCurrencyButton, withFlagButton, countryCode, containerButtonStyle, onOpen }) => {
    const withSomething = withCountryNameButton || withCallingCodeButton || withCurrencyButton;
    const { flagSizeButton: flagSize } = useTheme();
    return (React.createElement(TouchableOpacity, { activeOpacity: 0.7, onPress: onOpen },
        React.createElement(View, { style: [
                styles.container,
                withEmoji ? styles.containerWithEmoji : styles.containerWithoutEmoji,
                containerButtonStyle
            ] }, withSomething ? (React.createElement(FlagWithSomething, Object.assign({}, {
            countryCode,
            withEmoji,
            withCountryNameButton,
            withCallingCodeButton,
            withCurrencyButton,
            withFlagButton,
            flagSize: flagSize
        }))) : (React.createElement(Flag, Object.assign({}, { countryCode, withEmoji, withFlagButton, flagSize: flagSize }))))));
};
FlagButton.defaultProps = {
    withEmoji: Platform.OS === 'ios',
    withCountryNameButton: false,
    withCallingCodeButton: false,
    withCurrencyButton: false,
    withFlagButton: true
};
